cpp/cert/src/rules/CTR55-CPP/DoNotUseAnAdditiveOperatorOnAnIterator.ql

```diff
@@ -14,24 +14,65 @@
 import cpp
 import codingstandards.cpp.cert
 import codingstandards.cpp.Iterators
+import semmle.code.cpp.controlflow.Dominance
 
-from ContainerIteratorAccess it
+/**
+ * something like:
+ * `end = begin() + size()`
+ */
+Expr calculatedEndCheck(AdditiveOperatorFunctionCall calc) {
```

Comment:
```suggestion
Expr getComputedReferenceToOnePassedTheEndElement(Expr containerReference) {
```

I think we should clearly articulate what the predicates are doing to make the query more readable.

Note that the `result` and the parameter are swapped in my suggestion and the result is more general than `AdditiveOperationFunctionCall`, but we should still cast in the body for now until we support more patterns.

Reply:
completely agree with the clearer naming required! thanks!

along those lines, actually this doesnt return the call that provides the sufficient end check, the predicate says - there is such a check, and returns the qualifier of one of the important components of that, so that we can validate that it was for the same vector object as the iterator access comes from

ie we cant just swap the parameter and the result without a bit more of a rewrite,

lmk if this new naming /comments still doesnt make it clear enough though

Reply:
I only see a change in qldoc that is now inconsistent with the predicate.

The current implementation relates the qualifier of the calls to `begin` and `size` and their use in an `AdditivOperatorFunctionCall`. My suggestion just swaps that around where the qualifier is named `containerReference` and the operation representing the result of the end computation, the `AdditiveOperatorFunctionCall`, is now `result`. I think we can just swap those with a rewrite.

e.g.,

```ql
Expr getDerivedReferenceToOnePassedTheEndElement(Expr containerReference) {
  exists(
    ContainerAccessWithoutRangeCheck::ContainerSizeCall size,
    ContainerAccessWithoutRangeCheck::ContainerBeginCall begin,
    AdditiveOperatorFunctionCall calc
  |
    calc.getTarget().hasName("operator+")  and
    result = calc
  |
    DataFlow::localFlow(DataFlow::exprNode(size), DataFlow::exprNode(calc.getAChild*())) and
    DataFlow::localFlow(DataFlow::exprNode(begin), DataFlow::exprNode(calc.getAChild*())) and
    //make sure its the same container providing its size as giving the begin
    globalValueNumber(begin.getQualifier()) = globalValueNumber(size.getQualifier()) and
    containerReference = begin.getQualifier()
  )
}
```

Reply:
accepted! minus the get an arg part - in this its not as simple as an arg - the size access is a child - 
```
l = (i + std::min(static_cast<std::vector<int>::size_type>(10), v.size()));
```

and I know any child is not the most precise modelling of this, but I still think this is overall a lot more precise than before